Unify BOLT ports

cleishm opened the pull request:
Unifies the socket and websocket transports, and TLS variants, so they can be handled over a shared channel (port). Introduces a ternary value of `optional` for `dbms.bolt.tls.enabled`, which allows the server to accept both TLS and non-TLS connections.

The second commit in this PR extends the latter such that the server always accepts non-TLS connections, yet responds with a client error if the server requires communication over TLS (`dbms.bolt.tls.enabled=true`). This provides a more obvious error than an SSL negotiation failure.

Extends the integration tests to handle all 4 variations, and also test the TLS-required case.

jakewins commented:
This is a nice illustration of how the proposal you laid out could be implemented. Taking this as input for that discussion, I'm going to close this PR for now since there's not been a decision to change the design in this way.

cleishm committed:
Unify BOLT protocols onto a single port

cleishm committed:
Move TLS requirement to be handled within the protocol

This allows an informative message to be returned,
instructing the user that a TLS connection is required.

cleishm commented:
I can't keep this rebased and running in CI with it closed. We have a `NOT READY FOR MERGE` label that we apply whilst we're still discussing a proposal. I'll mark it as such now.

jakewins commented:
No need to keep the code up to date, this was straight forward when we tried it last year, and as you've illustrated here it remains so. If the team decides to adopt the design we will bring the code up to speed.

I'm going to close this now, again. If you disagree with that, please ping me privately.
